nexus_keycode/protocols/utils.py

```diff
@@ -58,3 +58,14 @@ def chunk(iteration):
     output_bits = bitstring.Bits().join(chunk(i) for i in range(iterations))
 
     return output_bits[:output_len]
+
+
+def generate_mac(input_val, secret_key):
+    """
+    :type input_val: 'str'
+    :type secret_key: 'str'
+    """
+    key_bytes = secret_key.encode("utf-8")
```

Comment:
This likely *works*, but the neither the secret key nor the input data should be `utf-8`. We expect to provide a secret key that is a raw, 16-byte value (e.g. `b"\x01\x02\x03... etc`).

By encoding it as `utf-8`, you are telling Python 'treat these bytes as representing a utf-8 string', which means 'when you see bytes in the string that represent UTF-8 codepoints, display the relevant character'. Heres an example:

```
>>> key = "\x01\x02\x03\x21\xc2\xb0"
>>> # \x01\x02\x03 don't have special meaning in UTF-8. \x21 is "!" in both ASCII and UTF-8. '\xc2\xb0' is "degree sign" in UTF-8
...
>>> encoded_key = key.encode('utf-8')
>>> encoded_key
b'\x01\x02\x03!\xc3\x82\xc2\xb0'
>>> encoded_key.decode('utf-8')
'\x01\x02\x03!Â°'
>>> # notice that python now shows "Â°, thats unexpected
... 
>>> decoded_key = encoded_key.decode('utf-8')
>>> str(decoded_key)
'\x01\x02\x03!Â°'
```

In other words, you *don't* want an encoding, you want raw bytes here. These bytes don't represent UTF-8 coded text.

Reply:
Concrete suggestion, though:

```
def generate_mac(input_val, secret_key):
    """
    :type input_val: 'bytes'
    :type secret_key: 'bytes'
    """
    function = siphash.SipHash_2_4(key_bytes, int_to_bytes(input_bytes))
    return u"{:06d}".format(function.hash() & 0xFFFFFFFF)[-6:]
```

Reply:
Note: This set of comments is optional, the code you have now works, but if we know we aren't dealing with a UTF-8 string at any point along the way, and the data is always bytestrings, it seems clearer to just skip the `encode` step (which *will* attempt to turn the data into a bytestring), and make sure the data being handled by `generate_mac` is a bytestring to begin with (rather than trying to convert it inside the function).